package pyqtc.pb;

message Position {
  optional string filename = 1;
  optional int32 line = 2;
  optional int32 column = 3;
}

message Type {
  message Reference {
    enum Kind {
      ABSOLUTE_TYPE_ID = 0;
      SCOPE_LOOKUP = 1;
      SCOPE_CALL = 2;
    }

    optional Kind kind = 1;
    optional string name = 2;
  }

  repeated Reference reference = 1;
}

message Scope {
  enum Kind {
    MODULE = 0;
    CLASS = 1;
    FUNCTION = 2;
  }

  optional Position declaration_pos = 1;
  optional string name = 2;
  optional Kind kind = 3;

  repeated Scope child_scope = 4;
  repeated Variable child_variable = 5;

  repeated Type base_type = 6;
  optional Type call_type = 7;
}

message Variable {
  enum Kind {
    FUNCTION_ARGUMENT = 0;
    SCOPE_VARIABLE = 1;
    MODULE_REF = 2;
  }

  optional Position declaration_pos = 1;
  optional string name = 2;
  optional Kind kind = 3;

  optional Type type = 4;
}
